SMTP_handler.py

```diff
@@ -0,0 +1,515 @@
+"""
+Enable SMTP access for aliases.
+
+This will allow a MUA to send directly from the alias without the need for the user to create a reverse-alias.
+
+"""
+
+import argparse
+import time
+import ssl
+import uuid
+import email
+
+import newrelic.agent
+
+from aiosmtpd.controller import Controller
+from aiosmtpd.smtp import AuthResult, LoginPassword, Envelope
+
+from sqlalchemy.orm.exc import ObjectDeletedError
+from email.message import Message
+from email.utils import formataddr, formatdate
+from init_app import load_pgp_public_keys
+from server import create_light_app
+from app.db import Session
+from app.log import LOG, set_message_id
+from app.email import status, headers
+from app.utils import sanitize_email
+from app.email.spam import get_spam_score
+from app.pgp_utils import PGPException
+from app.models import Alias, SMTPCredentials, EmailLog, Contact, Mailbox
+from app.config import (
+    SMTP_SSL_KEY_FILEPATH,
+    SMTP_SSL_CERT_FILEPATH,
+    NOREPLY,
+    ENABLE_SPAM_ASSASSIN,
+    SPAMASSASSIN_HOST,
+    MAX_REPLY_PHASE_SPAM_SCORE,
+    BOUNCE_PREFIX_FOR_REPLY_PHASE,
+    LOAD_PGP_EMAIL_HANDLER,
+)
+from app.email_utils import (
+    copy,
+    save_email_for_debugging,
+    sanitize_header,
+    get_header_unicode,
+    parse_full_address,
+    is_valid_alias_address_domain,
+    get_spam_info,
+    delete_all_headers_except,
+    sl_sendmail,
+    add_or_replace_header,
+    should_add_dkim_signature,
+    add_dkim_signature,
+    send_email,
+    render,
+    get_email_domain_part,
+    should_ignore_bounce,
+)
+
+from email_handler import (
+    send_no_reply_response,
+    get_or_create_contact,
+    handle_spam,
+    prepare_pgp_message,
+    replace_original_message_id,
+)
+
+
+def handle_SMTP(envelope, msg: Message, rcpt_to: str) -> (bool, str):
+    """
+    Return whether an email has been delivered and
+    the smtp status ("250 Message accepted", "550 Non-existent email address", etc)
+    """
+    website_email = rcpt_to
+
+    alias_address: str = envelope.mail_from
+    alias = Alias.get_by(email=alias_address)
+    if not alias:
+        LOG.e("Alias: %s isn't known", alias_address)
+        return False, status.E503
+
+    user = alias.user
+    if user.disabled:
+        LOG.e(
+            "User %s disabled, disable sending emails from %s to %s",
+            user,
+            alias,
+            website_email,
+        )
+        return [(False, status.E504)]
+
+    alias_domain = get_email_domain_part(alias_address)
+
+    # Sanity check: verify alias domain is managed by SimpleLogin
+    # scenario: a user have removed a domain but due to a bug, the aliases are still there
+    if not is_valid_alias_address_domain(alias.email):
+        LOG.e("%s domain isn't known", alias)
+        return False, status.E503
+
+    contact = Contact.get_by(website_email=website_email)
```

Comment:
As mentioned in another comment, I think we can also handle the case where website_email is a reverse alias here. This should be relatively easy IMO.

Reply:
I have added some code to handle this case.